service/storage_service.cc

```diff
@@ -5404,8 +5408,19 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
                     auto existing_node = x.first;
                     auto replacing_node = x.second;
                     slogger.info("replace[{}]: Added replacing_node={} to replace existing_node={}, coordinator={}", req.ops_uuid, replacing_node, existing_node, coordinator);
-                    tmptr->update_topology(replacing_node, get_dc_rack_for(replacing_node), locator::node::state::replacing);
+                    const auto dc_rack = get_dc_rack_for(replacing_node);
+                    tmptr->update_topology(replacing_node, dc_rack, locator::node::state::replacing);
                     tmptr->add_replacing_endpoint(existing_node, replacing_node);
+
+                    // topology cannot hold two IPs with different host ids, so
+                    // when replacing we must advertise the replaced_id for the ip, otherwise
+                    // topology will complain about host id of a local node changing and fail.
```

Comment:
"topology will complain" is not a correctness argument. It only means that the current state of the code is as it is. There seem to be multiple ways out of this situation (hypotheses to be tested):
- modify topology to allow multiple host IDs to have the same IP
- remove the complaining (allow topology to change host ID of given IP)
- the way that you picked

The question is -- given the possible behaviors of the system, which way is the correct one and why?
How is locator::topology used, and during replace operation, when someone calls get_host_id (for example) why is it (or isn't) correct to provide the replaced_id (and not the replacing_id, or perhaps both and let the caller choose?)

Reply:
I elaborated in the commit messages/comments.